source/extensions/filters/http/aws_request_signing/aws_request_signing_filter.cc

```diff
@@ -20,29 +23,72 @@ Extensions::Common::Aws::Signer& FilterConfigImpl::signer() { return *signer_; }
 FilterStats& FilterConfigImpl::stats() { return stats_; }
 
 const std::string& FilterConfigImpl::hostRewrite() const { return host_rewrite_; }
+bool FilterConfigImpl::useUnsignedPayload() const { return use_unsigned_payload_; }
 
 FilterStats Filter::generateStats(const std::string& prefix, Stats::Scope& scope) {
   const std::string final_prefix = prefix + "aws_request_signing.";
   return {ALL_AWS_REQUEST_SIGNING_FILTER_STATS(POOL_COUNTER_PREFIX(scope, final_prefix))};
 }
 
-Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers, bool) {
+Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers, bool end_stream) {
   const auto& host_rewrite = config_->hostRewrite();
+  const bool use_unsigned_payload = config_->useUnsignedPayload();
+
   if (!host_rewrite.empty()) {
     headers.setHost(host_rewrite);
   }
 
+  if (!use_unsigned_payload && !end_stream) {
```

Comment:
I touched on this in the original commit message:

> canonical must include the hashed payload for most services. The prior
> behavior of using UNSIGNED-PAYLOAD is an exception to the rule, which
> select services like s3 support, since hashing the payload may be
> impractical if the payload is very large.

The existing behavior of the filter is that the signature is based on `''` unless the service name is `s3`, `es`, or `glacier` in which case the literal `UNSIGNED-PAYLOAD` is used. This behavior is not configurable, so I think we should frame the rest of the discussion as whether the existing behavior should be the default going forward. I argue that the default behavior should not be the existing behavior...

The current behavior is incompatible with nearly all AWS services. I can't find any evidence that the changes made in #12020 actually made any cases change from failing to passing, since it only added `es` and `glacier` to the list of services which use `UNSIGNED-PAYLOAD` to calculate the content hash, and neither of those services document that they support that signature variant.

In the version history I worded the notes giving the benefit of doubt to what the code expressed, despite not believing this to be true:

> Previously, requests were never buffered, which only produced correct signatures for requests without a body, or for requests to S3, ES or Glacier, which used the literal string ``UNSIGNED-PAYLOAD``

We can actually search the AWS SDK models to see which services can use `UNSIGNED-PAYLOAD`:

```
find aws-sdk-cpp/code-generation/api-descriptions -name '*.json' -exec grep -lF v4-unsigned-body "{}" \+
aws-sdk-cpp/code-generation/api-descriptions/runtime.lex.v2-2020-08-07.normal.json
aws-sdk-cpp/code-generation/api-descriptions/runtime.lex-2016-11-28.normal.json
aws-sdk-cpp/code-generation/api-descriptions/s3-2006-03-01.normal.json
aws-sdk-cpp/code-generation/api-descriptions/ebs-2019-11-02.normal.json
aws-sdk-cpp/code-generation/api-descriptions/mediastore-data-2017-09-01.normal.json
```

Of those 4, only `s3` is included in the filter's list to use UNSIGNED-PAYLOAD, so the other 3 are in the same boat as all other services - the filter is only compatible with requests without a body.

I've used the new stats to verify that the new behavior still signs requests without a body in `decodeHeaders`. So, the only type of requests which are known to have worked with the existing behavior and would be treated differently by the new default behavior are requests to S3 with a body (uploads).

For requests to S3 with a body, the new filter will buffer and sign the payload by default. The consequence of that change depends on the size of uploads being handled and the buffer configuration - impact could range from negligible, to higher memory usage, to 413 responses being returned. Of course, the old behavior is available for this use case via the new `use_unsigned_payload` config setting.

I did consider retaining the old behavior for s3, but the filter needs to support signing payloads for s3 because s3 bucket policies can be used to deny requests with unsigned payloads. If the default s3 behavior is to use unsigned payloads, there would need to be a separate config setting like `sign_s3_payloads`, which would greatly increase the complexity of evaluating the filter config.

It is impossible to know how prevalently the filter may be used to facilitate uploads to s3 and whether the buffer limit would affect those usages. But, I argue that this one case cannot be assumed to be significant enough to justify the default filter behavior being incompatible with all other AWS services; nor could it justify the more confusing config interface and more complex config evaluation. The version history notes clearly state the change and provide instructions on how to enable the old behavior.